p2p/src/authenticated/actors/tracker/actor.rs

```diff
@@ -221,44 +219,39 @@ impl<E: Spawner + Rng + Clock + GClock + Metrics, C: Scheme> Actor<E, C> {
 
             // Iterate over peer set and decrement counts
             for peer in set.order.keys() {
-                if let Some(address) = self.peers.get_mut(peer) {
-                    if address.decrement() {
+                if let Some(record) = self.peers.get_mut(peer) {
+                    if record.decrement() {
                         self.peers.remove(peer);
                     }
                 }
             }
         }
+
+        // Update metrics
+        self.metrics.tracked.set(self.peers.len() as i64);
+        let blocked_count = self.peers.values().filter(|r| r.is_blocked()).count();
+        self.metrics.blocked.set(blocked_count as i64);
     }
 
+    /// Returns a list of peers that we have a known address for and were able to successfully reserve.
     #[allow(clippy::type_complexity)]
     fn handle_dialable(&mut self) -> Vec<(C::PublicKey, SocketAddr, Reservation<E, C>)> {
-        // Collect unreserved peers
-        let available_peers: Vec<_> = self
+        // Collect peers with known addresses
+        let peers: Vec<(C::PublicKey, SocketAddr)> = self
             .peers
-            .keys()
-            .filter(|peer| !self.connections.contains(*peer))
-            .cloned()
+            .iter()
+            .filter_map(|(peer, record)| record.get_address().map(|addr| (peer.clone(), addr)))
             .collect();
 
-        // Iterate over available peers
-        let mut reserved = Vec::new();
-        for peer in available_peers {
-            // Reserve the connection
-            let reservation = match self.reserve(peer.clone()) {
-                Some(reservation) => reservation,
-                None => continue, // can happen if rate limited
-            };
-
-            // Grab address
-            if let Some(address) = self.peers.get(&peer).unwrap().get_address() {
-                reserved.push((peer, address, reservation));
-            };
-        }
-        reserved
+        // Return all peers that we got a reservation for
+        peers
```

Comment:
I almost think we could just put `reserved` into `record` (now that there is some indirection on `record`)?

Reply:
I agree it seems tempting. One downside is that you can have a peer actor still operating even after the peer is removed from any active sets. Once this happens, the peer record is also removed. This would also remove the reservation. The actor will be sent a `Kill` message the next time they request a bitvec, which could be in "a while".

In this case it is nice that the reservation lives on until the peer actor actually dies and releases the reservation. Otherwise it's technically possible for the tracker to give out two simultaneous reservations for the same peer public key if the peer gets removed then re-added before requesting a bitvec